utxoscanner.go

```diff
@@ -0,0 +1,415 @@
+package neutrino
+
+import (
+	"container/heap"
+	"sync"
+	"sync/atomic"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil"
+	"github.com/btcsuite/btcwallet/waddrmgr"
+)
+
+// getUtxoResult is a simple pair type holding a spend report and error.
+type getUtxoResult struct {
+	report *SpendReport
+	err    error
+}
+
+// GetUtxoRequest is a request to scan for OutPoint from the height BirthHeight.
+type GetUtxoRequest struct {
+	// OutPoint is the target outpoint to watch for spentness.
+	OutPoint *wire.OutPoint
+
+	// BirthHeight is the height at which we expect to find the original
+	// unspent outpoint. This is also the height used when starting the
+	// search for spends.
+	BirthHeight uint32
+
+	// resultChan either the spend report or error for this request.
+	resultChan chan *getUtxoResult
+
+	// result caches the first spend report or error returned for this
+	// request.
+	result *getUtxoResult
+
+	// mu ensures the first response delivered via resultChan is in fact
+	// what gets cached in result.
+	mu sync.Mutex
+
+	quit chan struct{}
+}
+
+// deliver tries to deliver the report or error to any subscribers. If
+// resultChan cannot accept a new update, this method will not block.
+func (r *GetUtxoRequest) deliver(report *SpendReport, err error) {
+	select {
+	case r.resultChan <- &getUtxoResult{report, err}:
+	default:
+		log.Warnf("duplicate getutxo result delivered for "+
+			"outpoint=%v, spend=%v, err=%v",
+			r.OutPoint, report, err)
+	}
+}
+
+// Result is callback returning either a spend report or an error.
+func (r *GetUtxoRequest) Result() (*SpendReport, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	select {
+	case result := <-r.resultChan:
+		// Cache the first result returned, in case we have multiple
+		// readers calling Result.
+		if r.result == nil {
```

Comment:
Shouldn't we also check the internal result variable at the top before attempting to select? Otherwise, a caller can just get stuck here (if they're late).

Reply:
I don’t think it’s necessary since an exclusive lock is held throughout by the first caller, released, then acquired by any potential second calls to `Result`

Reply:
So `resultChan` should be closed by the time the lock is released

Reply:
So `resultChan` should be closed by the time the lock is released